lib/linked_list.rb

```diff
@@ -12,120 +12,346 @@ def initialize(value, next_node = nil)
+      prevNode = nil
+      while current_node
+        if current_node.data == value
+          if current_node.next
+            # remove current node from link by assigning current node's next to previous node
+            if prevNode
+              prevNode.next = current_node.next
+              current_node = nil
+            else
+              @head = current_node.next
+              current_node = nil
+            end
+          else
+            # remove current node from link
+            prevNode.next = nil
+            current_node = nil
+          end
+        else
+          prevNode = current_node
+          if current_node.next
+            current_node = current_node.next
+          else
+            return nil
+          end
+        end
+      end
+    else
+      return nil
     end
-
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    def has_cycle
-      raise NotImplementedError
+  end
+  
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  def reverse
+    # this will break if fed a linked list with fewer than three nodes
+    if @head
+      current = @head
+      current_next = @head.next
+      current_next_next = current_next.next
+      
+      while current.next
+        if current_next.next
+          current_next_next = current_next.next
+          current_next.next = current
+          current = current_next
+          current_next = current_next_next
+        else
+          current_next.next = current
+          current = current_next
+          break
+        end
+      end
+      
+      @head.next = nil
+      @head = current
+    else
+      return nil
     end
-
-
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    def get_first
-      raise NotImplementedError
+  end
+  
+  
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  def find_middle_value
+    # UNTESTED
+    if @head
+      count = 0
+      end_node = @head
+      while end_node
+        end_node = end_node.next
+        count += 1
+      end
+      mid_point = count/2
+      mid_node = @head
+      while count >= mid_point
+        mid_node = mid_node.next
+        count -= 1
+      end
+      return mid_node.data
+      
+    else
+      return nil
     end
-
-    # method that inserts a given value as a new last node in the linked list
-    def add_last(value)
-      raise NotImplementedError
+    
+  end
+  
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  def find_nth_from_end(n)
+    if @head
+      count = 0
+      current = @head
+      
+      # moves pointer n nodes along the list
+      while count < n
+        current = current.next

```

Comment:
What about if current is nil?

```suggestion
        return nil if current.nil?
        current = current.next
```